README.rst
==========
AppMetrica
==========

Application for integrating with Yandex AppMetrica https://appmetrica.yandex.ru/

Yandex Docs https://tech.yandex.ru/appmetrica/doc/mobile-api/push/use-cases-docpage/

.. image:: https://travis-ci.org/MyBook/appmetrica.svg?branch=master
    :target: https://travis-ci.org/MyBook/appmetrica
.. image:: https://codecov.io/gh/MyBook/appmetrica/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/MyBook/appmetrica

Before using API it is necessary:

1. Get `application id` from your account in appmetrica.yandex.ru

2. [Generate access token](https://tech.yandex.ru/appmetrica/doc/mobile-api/intro/authorization-docpage/)

Send push
---------

1. Create `API` instance::

    from appmetrica.push.api import PushAPI

    api = PushAPI(application_id, access_token)

2. Create group to combine the sending in the report::

    group_id = api.create_group('test-push-1', send_rate=500)

3. Pass `group_id`, `device list` and `message` to send_push method and call::

    from appmetrica.push.api import TokenTypes

    transfer_id = api.send_push(group_id, devices=devices, ios_message=ios_message,
                                android_message=android_message, tag='harry potter')

    devices - list of token objects like:
        [
            {
                "id_type": TokenTypes.APPMETRICA_DEVICE_ID,
                "id_values": ["123456789", "42"]
            },
            {
                "id_type": TokenTypes.IOS_IFA,
                "id_values": ["8A690667-6204-4A6A-9B38-85DE016....."]
            },
            {
                "id_type": TokenTypes.ANDROID_PUSH_TOKEN,
                "id_values": ["eFfxdO7uCMw:APA91bF1tN3X3BAbiJXsQhk-..."]
            }
        ]

    ios_message - push message for ios devices
    android_message - push message for android devices:
        {
            "silent": false,
            "content": {
                "title": "string",
                "text": "string",
                "sound": "disable",
                "data": "string"
            }
        }

4. To check status of push call `check_status` method::

    status = api.check_status(transfer_id)


List of available groups
------------------------

1. Create `API` instance::

    from appmetrica.push.api import PushAPI

    api = PushAPI(application_id, access_token)

2. Get list of groups

    group_id = api.get_groups()


Export tokens
-------------

1. Create `API` instance::

    from appmetrica.export.api import ExportAPI

    api = ExportAPI(application_id, access_token)

2. Call push_tokens method with necessary fields::

    data = api.export_push_tokens('token', 'ios_ifa', 'google_aid')


Export devices
--------------

1. Create `API` instance::

    from appmetrica.export.api import ExportAPI

    api = ExportAPI(application_id, access_token)

2. Call push_tokens method with necessary fields::

    date_till = datetime.now()
    date_from = date_till - timedelta(days=7)

    data = api.export_installations('ios_ifv', date_from=date_from, date_till=date_till)



Publish a release on PyPi
-------------------------

Install `twine <https://pypi.org/project/twine/>`_ globally::

  pip install twine

1. Don't forget to bump the package version::

    __version__ = '1.0.5'

2. Build the release::

    python setup.py sdist bdist_wheel

3. Publish the release on PyPi::

    twine upload dist/*


CHANGELOG.rst
Changelog
---------


1.0.6 (2020-11-11)
------------------
* Add send rate when create appmetrica group


1.0.5 (2020-10-01)
------------------
* Add retry for appmetrica request and declare python 3.8 support


1.0.4 (2020-06-26)
------------------
* Allow to send raw data


1.0.3 (2019-08-04)
------------------
* Declare python 3.7 support


1.0.2 (2019-01-09)
------------------
* Increased request timeout to 30 seconds


1.0.1 (2018-12-17)
------------------
* Replace deprecated `send` method to `send-batch`


1.0.0 (2017-12-14)
------------------
* Initial release


appmetrica/push/api.py
# coding: utf-8
import datetime
import json
import logging

from appmetrica.base import BaseAPI
from appmetrica.push import exceptions

logger = logging.getLogger(__name__)

MAX_NUMBER_IN_BATCH = 250000
MAX_NUMBER_OF_GROUPS = 5


class TokenTypes(object):
    APPMETRICA_DEVICE_ID = 'appmetrica_device_id'
    IOS_IFA = 'ios_ifa'
    GOOGLE_AID = 'google_aid'
    ANDROID_PUSH_TOKEN = 'android_push_token'
    IOS_PUSH_TOKEN = 'ios_push_token'


class PushAPI(BaseAPI):
    base_url = 'https://push.api.appmetrica.yandex.net/push/v1/'

    def create_group(self, name, send_rate=None):
        """
        Create group to combine the sending in the report
        :param name: Unique name of the group
        :param send_rate: Limit on the max speed of sending push messages (number per second)
        :return: Identifier of the created group
        """
        assert name
        group = {
            'app_id': self.app_id,
            'name': name,
        }
        if send_rate is not None:
            if not (100 <= send_rate <= 5000):
                logger.error('create_group with send_rate = %s failed. possible values from 100 to 5000', send_rate)
                raise exceptions.AppMetricaCreateGroupError
            group['send_rate'] = send_rate

        try:
            response_data = self._request('post', 'management/groups', json={'group': group})
        except Exception as exc:
            logger.error('create_group request with json %s failed due to %s', group, exc, exc_info=True)
            raise exceptions.AppMetricaCreateGroupError

        # response_data contains dict like:
        # {
        #   "group": {
        #     "id": XXXXXX,
        #     "app_id": XXXXXX,
        #     "name": "the_name_of_the_group",
        #     "send_rate": 5000,
        #   }
        # }
        return response_data['group']['id']

    def get_groups(self):
        """
        List of all available groups
        :return: List of dict with groups metadata
        """
        try:
            response_data = self._request('get', 'management/groups', params={'app_id': self.app_id})
        except Exception as exc:
            logger.error('get_groups request failed due to %s', exc, exc_info=True)
            raise exceptions.AppMetricaGetGroupsError

        # response_data contains dict like:
        # {
        #   "groups": [
        #     {
        #       "id": 1,
        #       "app_id": 3,
        #       "name": "name of group"
        #     },
        #     {
        #       "id": 2,
        #       "app_id": 3,
        #       "name": "name of other group"
        #     }
        #   ]
        # }
        return response_data['groups']

    def send_push(self, group_id, devices=None, ios_message=None, android_message=None, tag=None):
        """
        Sends push messages to the specified devices
        :param group_id: group to combine the sending in the report
        :param devices: list of token objects like:
            [
                {
                    "id_type": "appmetrica_device_id",
                    "id_values": ["123456789", "42"]
                },
                {
                    "id_type": "ios_ifa",
                    "id_values": ["8A690667-6204-4A6A-9B38-85DE016....."]
                },
                {
                    "id_type": "android_push_token",
                    "id_values": ["eFfxdO7uCMw:APA91bF1tN3X3BAbiJXsQhk-..."]
                }
            ]
        :param ios_message: push message for ios devices. dict like:
            {
                "silent": false,
                "content": {
                    "title": "string",
                    "text": "string",
                    "sound": "disable",
                    "data": "string"
                }
            }
        :param android_message: similarly ios_message param
        :param tag: send tag to combine the sending in the report
        :return: Identifier of the sending push (transfer_id)
        """
        if not devices:
            logger.error('send push error: devices are not provided')
            raise exceptions.AppMetricaSendPushError('devices are not provided')
        if not ios_message and not android_message:
            logger.error('send push error: messages are not provided')
            raise exceptions.AppMetricaSendPushError('messages are not provided')

        messages = {}
        if ios_message:
            messages['iOS'] = ios_message
        if android_message:
            messages['android'] = android_message

        tag = tag or datetime.datetime.now().isoformat()  # default tag
        data = {
            'push_batch_request': {
                'group_id': group_id,
                'tag': tag,
                'batch': [{
                    'messages': messages,
                    'devices': devices
                }]
            }
        }
        return self.send(data)

    def send(self, data):
        """
        Send batch of push messages

        :param data: dict with data, see https://appmetrica.yandex.ru/docs/mobile-api/push/post-send-batch.html
        :return: Identifier of the sending push (transfer_id)
        """
        self._validate(data)
        try:
            response_data = self._request('post', 'send-batch', json=data)
        except Exception as exc:
            logger.error('send_push request with json %s failed due to %s', data, exc, exc_info=True)
            raise exceptions.AppMetricaSendPushError
        return response_data['push_response']['transfer_id']

    def check_status(self, transfer_id):
        """
        Check status of sending pushes
        raise AppMetricaCheckStatusError if got an error

        :param transfer_id: Push identifier (return after call `send_push`)
        :return: status of push:
            failed — validation error. See `errors` field for detail
            in_progress — validation success. Sending messages in process
            pending — the request is accepted and waits for validation
            sent — sending completed
        """
        endpoint = 'status/{transfer_id}'.format(transfer_id=transfer_id)

        try:
            response_data = self._request('get', endpoint)
        except Exception as exc:
            logger.error('check_status request for transfer_id %s failed due to %s',
                         transfer_id, exc, exc_info=True)
            raise exceptions.AppMetricaCheckStatusError

        # response_data contain dict like:
        # {
        #   "transfer": {
        #     "creation_date": "2017-11-03T18:29:25+03:00",
        #     "id": XXXXXX,
        #     "status": "failed",
        #     "tag": "some_tag",
        #     "group_id": XXXXXX,
        #     "errors": [
        #       "Invalid push credentials for platform android"
        #     ]
        #   }
        # }
        if response_data['transfer']['status'] == 'failed':
            errors = response_data['transfer']['errors']
            if errors:
                raise exceptions.AppMetricaCheckStatusError(errors[0])

        return response_data['transfer']['status']

    @staticmethod
    def build_ios_message(**kwargs):
        """
        Create message object for ios platform
        :param kwargs:
            title (required) - subject of push
            text (required) - body of push
            data (optional) - extra data for client
        :return: Message object (dict)
        """
        # A message that is processed by the application in the background without displaying to the user.
        # Using silent push, you can send additional data, or check the delivery of notifications.
        silent = kwargs.get('silent') or False

        content = {
            'title': kwargs['title'],
            'text': kwargs['text'],
            'sound': kwargs.get('sound') or 'disable',
        }
        if 'extra_data' in kwargs:
            content['data'] = json.dumps(kwargs['extra_data'])

        return {
            "silent": silent,
            "content": content,
        }

    def _validate(self, data):
        required_fileds = ['group_id', 'tag', 'batch']
        for field in required_fileds:
            if field not in data['push_batch_request']:
                raise exceptions.AppMetricaSendPushError('field %s is required' % field)

        batch = data['push_batch_request']['batch']
        if not batch:
            logger.error('send push error: batch cant be empty')
            raise exceptions.AppMetricaSendPushError('batch cant be empty')

        try:
            num_devices = 0
            for messages in batch:
                device_groups = messages['devices']
                if len(device_groups) > MAX_NUMBER_OF_GROUPS:
                    logger.error('send push error: too many devices groups')
                    raise exceptions.AppMetricaSendPushError('too many devices groups')
                num_devices += sum([len(devices['id_values']) for devices in device_groups])
            if num_devices > MAX_NUMBER_IN_BATCH:
                logger.error('send push error: too many devices')
                raise exceptions.AppMetricaSendPushError('too many devices')
        except KeyError:
            raise exceptions.AppMetricaSendPushError('invalid format of data')
